File: ios/Sources/MicState.swift
import Combine
import Foundation

/// Household pause state, read from the recorder's `/api/capture`. Mirrors the
/// Android `CaptureState` — the API is the authority on pause, never the socket.
/// `reachable` is false when the API call failed (off the LAN), so the banner hides.
///
/// Spec-vs-status: `running`/`pausedUntil` is the mic's confirmed word, `desired*`
/// is the intent (moves the instant a button is pressed), `settled` says they
/// agree. Unsettled renders as "Pausing…"/"Resuming…" — never a flap between the
/// intent just set and a not-yet-caught-up report. Defaults read an older server's
/// confirmed-only answer as settled.
struct CaptureState: Equatable {
    var running: Bool
    var reachable: Bool
    var pausedUntil: Date?
    var desiredRunning = true
    var desiredPausedUntil: Date?
    var settled = true
    var micReachable = true
    /// Fingerprint echoed back as ?known= to long-poll /api/capture — the request
    /// hangs until the state changes. Nil on an older server (plain polling).
    var stateToken: String?
}

/// Connection / streaming phase, used to drive the status card text.
enum MicPhase: Equatable {
    case stopped
    case waitingForHost  // can't reach the recorder
    case paused  // recorder is up but household recording is paused
    case streaming  // connected and pumping PCM
}

/// Observable app state shared by the UI, the audio capture, and the stream client.
@MainActor
final class MicState: ObservableObject {
    @Published var running = false  // user pressed Start
    /// False while the audio engine will not open. Carried by the heartbeat so a
    /// running-but-deaf app SAYS so — before #887 it simply stopped beating, and the
    /// check went red for the wrong reason. Starts true: "not known to be broken".
    @Published var micOk = true
    @Published var connected = false  // TCP up and streaming
    @Published var phase: MicPhase = .stopped
    @Published var level: Float = 0  // 0...1 meter position
    /// Bytes of captured audio this app discarded because it could not deliver
    /// them (the spool overran). The phone is the ONLY place that knows: those
    /// samples never reach the network, so no server-side check can see them.
    /// Zero is the normal value; anything else is speech heard and lost.
    @Published var droppedBytes: Int = 0
    @Published var capture = CaptureState(running: true, reachable: false, pausedUntil: nil)
    @Published var sources: [SourceStatus] = []  // fleet liveness for the Devices panel

    var deviceID: String { Prefs.deviceID }

    var statusText: String {
        switch phase {
        case .stopped: return "Stopped"
        case .waitingForHost: return "Waiting for recall host"
        case .paused: return "Household recording paused"
        case .streaming: return "Streaming to \(Prefs.host)"
        }
    }
}

File: android/app/src/main/kotlin/org/recall/mic/MicState.kt
package org.recall.mic

import android.util.Log
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

// Shared tag for UI-state-change logging; filter with `adb logcat -s recall-ui:I`.
const val UI_LOG = "recall-ui"

/**
 * Live streaming state published by [StreamService] and observed by the UI. Same
 * process, so this is just shared in-memory state (StateFlow is thread-safe) — no
 * binding or IPC. The service writes from its capture thread; the UI collects it.
 */
object MicState {
    private val _running = MutableStateFlow(false)
    val running: StateFlow<Boolean> = _running.asStateFlow()

    private val _connected = MutableStateFlow(false)
    val connected: StateFlow<Boolean> = _connected.asStateFlow()

    /**
     * False while the audio engine will not open (permission revoked, mic held by
     * another app). Carried by the heartbeat so a running-but-deaf app SAYS so:
     * before #887 such an app simply stopped beating, and the check went red for
     * the wrong reason. Starts true — "not known to be broken".
     */
    private val _micOk = MutableStateFlow(true)
    val micOk: StateFlow<Boolean> = _micOk.asStateFlow()

    fun setMicOk(value: Boolean) {
        _micOk.value = value
    }

    /**
     * Bytes of captured audio this app discarded because it could not deliver them
     * — the spool overran (PcmSpool). The phone is the ONLY place that knows this
     * happened: the samples never reach the network, so no server-side check can
     * see them. Zero is the normal, expected value; anything else is speech that
     * was heard and lost, and it belongs in the heartbeat beside micOk.
     */
    private val _droppedBytes = MutableStateFlow(0L)
    val droppedBytes: StateFlow<Long> = _droppedBytes.asStateFlow()

    fun setDroppedBytes(value: Long) {
        _droppedBytes.value = value
    }

    /** Most recent mic peak amplitude, 0f..1f, for the level meter. */
    private val _level = MutableStateFlow(0f)
    val level: StateFlow<Float> = _level.asStateFlow()

    /**
     * The household capture (pause) state from /api/capture — the single value the
     * screen and the notification both render, so they can't disagree. Written by
     * whichever component is alive to poll (the open screen and/or the running
     * service); null until first read / when unreachable.
     */
    private val _capture = MutableStateFlow<CaptureState?>(null)
    val capture: StateFlow<CaptureState?> = _capture.asStateFlow()

    fun setCapture(value: CaptureState?) {
        val old = _capture.value
        if (value?.running != old?.running || value?.settled != old?.settled) {
            Log.i(
                UI_LOG,
                "household capture running=${value?.running} " +
                    "desired=${value?.desiredRunning} settled=${value?.settled}",
            )
        }
        _capture.value = value
    }

    fun setRunning(value: Boolean) {
        if (value != _running.value) Log.i(UI_LOG, "service running=$value")
        _running.value = value
    }

    fun setConnected(value: Boolean) {
        if (value != _connected.value) Log.i(UI_LOG, "stream connected=$value (drives 'Streaming')")
        _connected.value = value
        if (!value) _level.value = 0f
    }

    fun setLevel(value: Float) {
        _level.value = value
    }
}
